File: 1-easy/00977-squares-of-a-sorted-array/solution.cpp
class Solution {
public:
    vector<int> sortedSquares(vector<int>& nums) {
        int nums_len = nums.size();
        for (int i{0}; i < nums_len; ++i) {
            nums[i] *= nums[i];
        }

        int left = 0;
        int right = nums_len - 1;
        vector<int> res(nums_len, 0);
        int res_index = nums_len;

        while (left <= right) {
            res_index -= 1;
            if (nums[left] < nums[right]) {
                res[res_index] = nums[right];
                right -= 1;
            } else {
                res[res_index] = nums[left];
                left += 1;
            }
        }

        return  res;
    }
};

File: 1-easy/00977-squares-of-a-sorted-array/solution.rs
impl Solution {
    pub fn sorted_squares(mut nums: Vec<i32>) -> Vec<i32> {
        let nums_len = nums.len();
        for i in 0..nums_len {
            nums[i] *= nums[i];
        }

        let mut left = 0_usize;
        let mut right = nums_len - 1;
        let mut res = vec![0_i32; nums_len];
        let mut res_index = nums_len - 1;

        while left <= right {
            if nums[left] < nums[right] {
                res[res_index] = nums[right];
                right -= 1;
            } else {
                res[res_index] = nums[left];
                left += 1;
            }

            // This will overflow in last iteration, but makes no problem for current code
            res_index -= 1;
        }

        res
    }
}
